storageccl: make KeyRewriter lookup O(logn) and memoized

dt opened the pull request:
The previous code was O(n) in terms of number of tables, and could be very slow for very large numbers of tables.

The O(n) code on the two-prefix benchmark clocked in at about 10ns/op. Sorting the prefixes allows binary searching to find a matching one, but on this small benchmark, the binary search version alone actually takes closer to 30ns/op. 

Adding memoization should help considerably, both in the small benchmark, but also in real world usage too: as keys are passed to the rewriter in order and will almost always be matched by the same table prefix, memoization of just the last match likely has nearly 100% hit rate, and means that the log n lookup should only happen ~once per range.

dt committed:
storageccl: reduce key rewrite lookup from O(n) to O(logn)

This change keeps key rewriter's rewriters sorted so that we can use
binary search to lookup a rewriter instead of iterating the whole list.

Release note: none.

dt committed:
storageccl: memoize last matching rewrite rule

Release note: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/58045)
<!-- Reviewable:end -->

dt committed:
backupccl: reuse rewriter between processing spans

Previously before processing each span we would re-derive the KeyRewriter
from the original descriptors -- this is both expensive and means that
the new rewriter no longer has the old rewriter's cached prefix that
often -- for sequential spans -- would be useful.

Release note: none.

Release note (<category, see below>): <what> <show> <why>

miretskiy reviewed (approved):
> Reviewed 2 of 3 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pbardea)



<!-- Sent from Reviewable.io -->

dt commented:
TFTRs!

bors r+
